### ISSUE TITLE
CI/CD attempts to deploy from feature branch

### ISSUE THREAD

**Issue by gntpet** [microsoft/AL-Go#558]
Hey @freddydk,

I just noticed that CI/CD is trying to deploy to my environment from other branches than **main**
Check the screenshot bellow. As you can see I'm running CICD for a feature branch. To my understanding it should not event try to deploy from this branch. 

![image](https://github.com/microsoft/AL-Go/assets/16342128/ed7afee2-45c2-42b6-b9c8-af04e2e1c3ef)

**My environment setup looks like this:**
![image](https://github.com/microsoft/AL-Go/assets/16342128/8367ab9b-0209-4cc1-aad2-9f9bcc5270c9)

**my settings are:**
**AL-Go-Settings.json**
```
{
  "type": "PTE",
  "templateUrl": "https://github.com/microsoft/AL-Go-PTE@main",
  "treatTestFailuresAsWarnings": true,
  "doNotRunBcptTests": true,
  "enableCodeCop": true,
  "rulesetFile": "../al.ruleset.json",
  "companyName": "TNM Automated Test"
}
```

**settings.json** of my project, which is configured for the environment. 
```
{
    "country": "it",
    "artifact": "bcartifacts/sandbox/22.0/it/latest",
    "appFolders": [],
    "testFolders": [],
    "appDependencyProbingPaths": [
        {
            "repo": ".",
            "release_status": "include",
            "projects": "SRS.core"
        }
    ],
    "installApps": [
        "https://verysecreturl"
    ]
}
```
Is it a bug, or am I'm missing some setup?

Best Regards, 
Gintautas

**Comment by freddydk** [comment]
It should default to only deploy from main, but you should be able to setup from which branches an environment will deploy.
Will investigate this one.

**Comment by gntpet** [comment]
hey @freddydk, do you have any update on this ticket?

**Comment by rbability** [comment]
Did you try to set a
`
"CICDPushBranches": [
    "main",
    "otherbranch"
  ],
`

in your AL-Go-Settings.json? We put in all the branches which should trigger a CI/CD run and it works fine here with the latest AL-Go version.
Do not forget to update AL-Go after changing this setting, so your cicd.yml triggers get updated with your branches.

**Comment by freddydk** [comment]
@rbability - the issue for @gntpet is not that the CI/CD workflow doesn't run - it is that it also triggers the deploy mechanism.

**Comment by rbability** [comment]
@freddydk I totally misunderstood, sorry. I only recently started using AL-Go seriously and I thought that a CI/CD run picks up every configured environment (except the ones with a (Production) affix) was by design.
We utilized this behaviour now to update the BC Test environment automatically whenever something is pushed to the Test branch of the Repo. That´s quite comfortable.

And to rollout code from the Main branch (which for my projects currently is only fully tested and approved code) we use the Create a Release + Publish to Environment actions to rollout to the Production environment.

If this is not the intended behaviour and CI/CD should only use the main branch, it would be nice to have the option to somehow make it possible to include other branches / environments to keep the behaviour I mentioned above.

**Comment by freddydk** [comment]
By default should only the main branch deploy to QA environments, but you should be able to set on the environment which branches deploys to that.

**Comment by rbability** [comment]
As far as I understood, the Deployment branches in the environment you set limit which branches are allowed to utilize this environment for deployments. If I understood you corretly, CI/CDs default behaviour should only start the deployment step when it runs on the main branch and a CI/CD run on a test branch would never trigger the deployment step and so environments that "could" be used by test would not matter at all.

That's not an issue for me. I am still learning. And if the current behaviour is a bug, I'd better not utilize it ;-)
I try to stick to how it is intended to work to have the least problems in the future with updates.

**Comment by freddydk** [comment]
No, the idea is that by defrault - main is the only branch - unless you specify differently on the environment (either in AL-Go Settings or in GitHub Environment Settings)

**Comment by freddydk** [comment]
The problem here was, that if the environment was defined in GitHub, then AL-Go would assume that branch policies was defined on the environment and select all branches itself - waiting for GitHub to reject the branch.
The problem with this approach is that if you have a Teams or Enterprise SKU, then the environment will be auto-created when deploying the first time (even if defined in settings).

In all cases, if you don't define branch policies in GitHub Environments (and you have a Teams or Enterprise SKU), then AL-Go currently will try to deploy from feature branches as well.

The fix for this is to only rely on GitHub branch policies if they actually are defined - else use the default branch policies.

**Comment by rbability** [comment]
Relying only on the Branch policies sounds very good and would be very intuitive to control where a deployment should run and where not.
If I remember correctly, currently when a CI/CD run is blocked to deploy (due to having branch policies set on the environment) when running on a specific branch, the Deployment Step will show an error and I think then the whole CI/CD run was flagged with an error. So this behaviour will change too, right?

**Comment by freddydk** [comment]
Yes

**Comment by gntpet** [comment]
hey @freddydk, 
do you have any date when the fix will be released? as far as i understand, it's still on preview

**Comment by freddydk** [comment]
Shipped in v3.2
